File: DSA/src/Questions/Greedy/Candy.java
package Questions.Greedy;

/*
https://leetcode.com/problems/candy/description/
There are n children standing in a line. Each child is assigned a rating value given in the integer array ratings.

You are giving candies to these children subjected to the following requirements:

Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.
Return the minimum number of candies you need to have to distribute the candies to the children.

 

Example 1:

Input: ratings = [1,0,2]
Output: 5
Explanation: You can allocate to the first, second and third child with 2, 1, 2 candies respectively.
Example 2:

Input: ratings = [1,2,2]
Output: 4
Explanation: You can allocate to the first, second and third child with 1, 2, 1 candies respectively.
The third child gets 1 candy because it satisfies the above two conditions.
 
Time Complexity: O(n)
Space Complexity: O(n)
 */
public class Candy {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int[] candies = new int[n];

        // Step 1: Initialize candies
        for (int i = 0; i < n; i++) {
            candies[i] = 1;
        }

        // Step 2: Left to Right pass
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) {
                candies[i] = candies[i - 1] + 1;
            }
        }

        // Step 3: Right to Left pass
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                candies[i] = Math.max(candies[i], candies[i + 1] + 1);
            }
        }

        // Step 4: Calculate the total candies
        int totalCandies = 0;
        for (int candy : candies) {
            totalCandies += candy;
        }

        return totalCandies;
    }

    public static void main(String[] args) {
        Candy candy = new Candy();
        int[] ratings = { 1, 0, 2 };
        System.out.println(candy.candy(ratings)); // Output: 5
    }
}

File: DSA/Python/Questions/Greedy/Candy.py
"""
# Question: Candy Distribution
# Link: https://leetcode.com/problems/candy/

# Time Complexity: O(n)
# Space Complexity: O(n)

# Algorithm:
# 1. Initialize candy array with 1 candy each
# 2. Forward pass: compare with left neighbor
# 3. Backward pass: compare with right neighbor
# 4. Return total candies needed
"""


class Candy:
    def candy(self, ratings: list[int]) -> int:
        n = len(ratings)
        candies = [1] * n

        # Forward pass
        for i in range(1, n):
            if ratings[i] > ratings[i - 1]:
                candies[i] = candies[i - 1] + 1

        # Backward pass
        for i in range(n - 2, -1, -1):
            if ratings[i] > ratings[i + 1]:
                candies[i] = max(candies[i], candies[i + 1] + 1)

        return sum(candies)


def main():
    solution = Candy()
    ratings = [1, 0, 2]
    print(solution.candy(ratings))  # Expected: 5


if __name__ == "__main__":
    main()
